### PR TITLE
Publicize the number of workspaces per state as a prometheus metric.

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
This adds the number of workspaces in different states as a Prometheus metric.

### What issues does this PR fix or reference?

#12092

#### Release Notes
* https://github.com/eclipse/che-docs/pull/656

### CODE REVIEW

**Comment by metlos** [comment]
Can core depend on wsmaster? Or do you suggest reimplementing parts of ws-activity in core-metrics?

**Comment by skabashnyuk** [comment]
Good question. What if we create module wsmaster/che-core-api-metrics/..../WorkspaceActivityMeterBinder.java that will reuse WorkspaceActivityManager.java methods?

**Review by mshaposhnik — APPROVED** [review]

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceActivityMeterBinder.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
+import org.eclipse.che.api.workspace.activity.WorkspaceActivityManager;
+
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceActivityMeterBinder.java:22]: Do we need a brief explanation of what is this class for?

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WsMasterMetricsModule.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.metrics;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.multibindings.Multibinder;
+import io.micrometer.core.instrument.binder.MeterBinder;
+
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WsMasterMetricsModule.java:17]: Do we need a brief explanation of what is this class for?

**Review by skabashnyuk — APPROVED** [review]
Overall looks good. Can you add Prometheus output in the description?

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
This is what the metrics look like in the output of `<CHE_SERVER>:8087/metrics` for Prometheus consumption:

```
# HELP che_workspace_status The number of workspaces in a given status
# TYPE che_workspace_status gauge
che_workspace_status{area="workspace",status="STARTING",} 2.0
che_workspace_status{area="workspace",status="RUNNING",} 12.0
che_workspace_status{area="workspace",status="STOPPING",} 0.0
che_workspace_status{area="workspace",status="STOPPED",} 6.0
```

File: wsmaster/che-core-api-workspace-activity/src/test/java/org/eclipse/che/api/workspace/activity/spi/tck/WorkspaceActivityDaoTest.java [hunk 3]
```diff
@@ -186,6 +186,25 @@ public void shouldFindActivityByLastStatusChangeTime(WorkspaceStatus status) thr
     assertEquals(found.getItems(), singletonList(activities[0].getWorkspaceId()));
   }
 
+  @Test(dataProvider = "allWorkspaceStatuses")
+  public void shouldCountWorkspacesInStatus(WorkspaceStatus status) throws Exception {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace-activity/src/test/java/org/eclipse/che/api/workspace/activity/spi/tck/WorkspaceActivityDaoTest.java]: I'm OK with this test but just want to note that mostly we write our tests in `GivenWhenThen` manner while this method invokes `countWorkspacesInStatus` three times and tests different cases in one method.

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceActivityMeterBinder.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  @Inject
+  public WorkspaceActivityMeterBinder(WorkspaceActivityManager activityManager) {
+    this.activityManager = activityManager;
+  }
+
+  @Override
+  public void bindTo(MeterRegistry registry) {
+    for (WorkspaceStatus s : WorkspaceStatus.values()) {
+      Gauge.builder("che.workspace.status", activityManager, __ -> count(s))
```
Comment (sleshchenko) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceActivityMeterBinder.java]: Do we really need put `activityManager` here and the ignore it by `ToDoubleFunction`? Can't we rewrite it with `Gauge#(String name, Supplier<Number> f)` like:
```java
      Gauge.builder("che.workspace.status", () -> count(s))
          .tag("status", s.name())
          .tag("area", "workspace")
          .description("The number of workspaces in a given status")
          .register(registry);
```
?

**Review by sleshchenko — APPROVED** [review]
LGTM

Please take a look my inline comments

**Comment by metlos** [comment]
ci-build

**Comment by metlos** [comment]
ci-test

**Pull request merged**